src/state/swap/hooks.ts

```diff
@@ -109,150 +97,212 @@ export function tryParseAmount(value?: string, currency?: Currency, chainId?: nu
 
-  const { trades: inTrades, loading: inLoading } = useEcoRouterExactIn(
-    isExactIn ? parsedAmount : undefined,
-    outputCurrency ?? undefined
+  // useCurrency and useToken returns a new object every time,
+  // so we need to compare the addresses as strings
+  const parsedAmountString = `${parsedAmount?.currency.address?.toString()}-${parsedAmount?.raw?.toString()}`
+  const recipientLookupComputed = `${recipientLookup.loading}-${recipientLookup?.address}-${recipientLookup?.name}`
+
+  useWhatChanged(
+    [
+      account,
+      useMultihops,
+      recipientLookupComputed,
+      chainId,
+      inputCurrency?.address,
+      outputCurrency?.address,
+      parsedAmountString,
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      relevantTokenBalances[0]?.raw.toString(),
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+      relevantTokenBalances[1]?.raw.toString(),
+      allowedSlippage,
+      recipient,
+      isExactIn,
+    ],
+    `account,useMultihops,recipientLookupComputed,chainId,inputCurrency?.address,outputCurrency?.address,parsedAmountString,relevantTokenBalances[0]?.raw.toString(),relevantTokenBalances[1]?.raw.toString(),allowedSlippage,recipient,isExactIn`
   )
-  const { trades: outTrades, loading: outLoading } = useEcoRouterExactOut(
-    inputCurrency ?? undefined,
-    !isExactIn ? parsedAmount : undefined
-  )
-  const bestTradeExactIn = inTrades[0]
-  const bestTradeExactOut = outTrades[0]
 
-  const allPlatformTrades = isExactIn ? inTrades : outTrades
-  // If overridden platform selection and a trade for that platform exists, use that.
-  // Otherwise, use the best trade
-  let platformTrade
-  if (platformOverride) {
-    platformTrade = allPlatformTrades?.filter(t => t?.platform === platformOverride)[0]
-  }
-  const trade = platformTrade ? platformTrade : isExactIn ? bestTradeExactIn : bestTradeExactOut
-
-  const currencyBalances = {
-    [Field.INPUT]: relevantTokenBalances[0],
-    [Field.OUTPUT]: relevantTokenBalances[1],
-  }
+  useEffect(() => {
+    const [inputCurrencyBalance] = relevantTokenBalances
 
-  const currencies: { [field in Field]?: Currency } = {
-    [Field.INPUT]: inputCurrency ?? undefined,
-    [Field.OUTPUT]: outputCurrency ?? undefined,
-  }
+    // Require a connected wallet
+    if (!account) {
+      setInputError(SWAP_INPUT_ERRORS.CONNECT_WALLET)
+      return
+    }
+    // Require two currencies to be selected
+    if (!inputCurrency || !outputCurrency) {
+      setInputError(SWAP_INPUT_ERRORS.SELECT_TOKEN)
+      return
+    }
+    // Require a valid input amount
+    if (!parsedAmount) {
+      setInputError(SWAP_INPUT_ERRORS.ENTER_AMOUNT)
+      return
+    }
+    // Require sufficient input balance
+    if (inputCurrencyBalance && JSBI.greaterThan(parsedAmount.raw, inputCurrencyBalance.raw as JSBI)) {
+      setInputError(SWAP_INPUT_ERRORS.INSUFFICIENT_FUNDS)
+      return
+    }
```

Comment:
Should we stop trade simulation even if one of these errors occurs? I think we can still get and show the best trades and then show some errors. Here is my suggestion 👇

```suggestion
    // Require a connected wallet
    if (!account) {
      setInputError(SWAP_INPUT_ERRORS.CONNECT_WALLET)
    }
    // Require two currencies to be selected
    if (!inputCurrency || !outputCurrency) {
      setInputError(SWAP_INPUT_ERRORS.SELECT_TOKEN)
      return
    }
    // Require a valid input amount
    if (!parsedAmount) {
      setInputError(SWAP_INPUT_ERRORS.ENTER_AMOUNT)
      return
    }
    // Require sufficient input balance
    if (inputCurrencyBalance && JSBI.greaterThan(parsedAmount.raw, inputCurrencyBalance.raw as JSBI)) {
      setInputError(SWAP_INPUT_ERRORS.INSUFFICIENT_FUNDS)
    }
```

Reply:
The goal of this PR is to save resources; the Eco Router consumes a lot. hence, the Eco Router should be triggered when the right parameters are set. It's not rational for users to request a quote from the Eco Router when they don't have sufficient funds in the first place :)